src/main/java/com/iota/iri/TransactionValidator.java

```diff
@@ -167,51 +170,46 @@ public void addSolidTransaction(Hash hash) {
     private Runnable spawnSolidTransactionsPropagation() {
         return () -> {
             while(!shuttingDown.get()) {
-                Set<Hash> newSolidHashes = new HashSet<>();
-                useFirst.set(!useFirst.get());
-                synchronized (cascadeSync) {
-                    if (useFirst.get()) {
-                        newSolidHashes.addAll(newSolidTransactionsTwo);
-                    } else {
-                        newSolidHashes.addAll(newSolidTransactionsOne);
-                    }
-                }
-                Iterator<Hash> cascadeIterator = newSolidHashes.iterator();
-                Set<Hash> hashesToCascade = new HashSet<>();
-                while(cascadeIterator.hasNext() && !shuttingDown.get()) {
-                    try {
-                        Hash hash = cascadeIterator.next();
-                        TransactionViewModel transaction = TransactionViewModel.fromHash(tangle, hash);
-                        Set<Hash> approvers = transaction.getApprovers(tangle).getHashes();
-                        for(Hash h: approvers) {
-                            TransactionViewModel tx = TransactionViewModel.fromHash(tangle, h);
-                            if(quietQuickSetSolid(tx)) {
-                                tx.update(tangle, "solid");
-                            } else {
-                                if (transaction.isSolid()) {
-                                    addSolidTransaction(hash);
-                                }
-                            }
-                        }
-                    } catch (Exception e) {
-                        log.error("Some error", e);
-                        // TODO: Do something, maybe, or do nothing.
-                    }
-                }
-                synchronized (cascadeSync) {
-                    if (useFirst.get()) {
-                        newSolidTransactionsTwo.clear();
-                    } else {
-                        newSolidTransactionsOne.clear();
+                propagateSolidTransactions();
+            }
+        };
+    }
+
+    void propagateSolidTransactions() {
+        Set<Hash> newSolidHashes = new HashSet<>();
+        useFirst.set(!useFirst.get());
+        //synchronized to make sure no one is changing the newSolidTransactions collections during addAll
+        synchronized (cascadeSync) {
+            if (useFirst.get()) {
+                newSolidHashes.addAll(newSolidTransactionsTwo);
+                newSolidTransactionsTwo.clear();
+            } else {
+                newSolidHashes.addAll(newSolidTransactionsOne);
+                newSolidTransactionsOne.clear();
+            }
+        }
+        Iterator<Hash> cascadeIterator = newSolidHashes.iterator();
+        while(cascadeIterator.hasNext() && !shuttingDown.get()) {
```

Comment:
Does it make sense to check this here by the way? We're checking in the Runnable anyway

Reply:
I think that this will allow for early exit from the loop.
Should it exist? not sure

I just don't want to add more changes for this PR. I want to do sanity tests and then improve it.